pkg/pusher/pusher_test.go

```diff
@@ -214,6 +234,107 @@ func TestSendChunkAndTimeoutinReceivingReceipt(t *testing.T) {
 	}
 }
 
+func TestPusherClose(t *testing.T) {
+	// create a trigger  and a closestpeer
+	triggerPeer := swarm.MustParseHexAddress("6000000000000000000000000000000000000000000000000000000000000000")
+	closestPeer := swarm.MustParseHexAddress("f000000000000000000000000000000000000000000000000000000000000000")
+
+	var (
+		goFuncStartedC    = make(chan struct{})
+		pusherClosedC     = make(chan struct{})
+		goFuncAfterCloseC = make(chan struct{})
+	)
+
+	defer func() {
+		close(goFuncStartedC)
+		close(pusherClosedC)
+		close(goFuncAfterCloseC)
+	}()
+
+	pushSyncService := pushsyncmock.New(func(ctx context.Context, chunk swarm.Chunk) (*pushsync.Receipt, error) {
+		goFuncStartedC <- struct{}{}
+		<-goFuncAfterCloseC
+		return nil, nil
+	})
+
+	_, p, storer := createPusher(t, triggerPeer, pushSyncService, mock.WithClosestPeer(closestPeer))
+
+	chunk := createChunk()
+
+	_, err := storer.Put(context.Background(), storage.ModePutUpload, chunk)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	storer.modeSetMu.Lock()
+	if storer.closed == true {
+		t.Fatal("store should not be closed")
+	}
+	if storer.setBeforeCloseCount != 0 {
+		t.Fatalf("store 'Set' called %d times before close, expected 0", storer.setBeforeCloseCount)
+	}
+	if storer.setAfterCloseCount != 0 {
+		t.Fatalf("store 'Set' called %d times after close, expected 0", storer.setAfterCloseCount)
+	}
+	storer.modeSetMu.Unlock()
+
+	select {
+	case <-goFuncStartedC:
+	case <-time.After(5 * time.Second):
+		t.Fatal("timed out waiting to start worker job")
+	}
+
+	// close in the background
+	go func() {
+		p.Close()
+		storer.Close()
+		pusherClosedC <- struct{}{}
+	}()
+
+	select {
+	case <-pusherClosedC:
+	case <-time.After(2 * time.Second):
+		// internal 5 second timeout that waits for all pending push operations to terminate
+	}
+
+	storer.modeSetMu.Lock()
+	if storer.setBeforeCloseCount != 0 {
+		t.Fatalf("store 'Set' called %d times before close, expected 0", storer.setBeforeCloseCount)
+	}
+	if storer.setAfterCloseCount != 0 {
+		t.Fatalf("store 'Set' called %d times after close, expected 0", storer.setAfterCloseCount)
+	}
+	storer.modeSetMu.Unlock()
+
+	select {
+	case goFuncAfterCloseC <- struct{}{}:
+	case <-time.After(5 * time.Second):
+		t.Fatal("timed out waiting for chunk")
+	}
+
+	// we need this to allow some goroutines to complete
+	time.Sleep(10 * time.Millisecond)
```

Comment:
This sleep can be longer.

Reply:
My bad, forgot to push latest local change.